File: worker/akapela_worker/audio.py
"""Thin wrappers over ffmpeg and ffprobe.

Every stored audio master is 44.1 kHz stereo WAV (ADR 0005).
"""

from __future__ import annotations

import json
import re
import subprocess
from pathlib import Path

BACKING_SAMPLE_RATE = 44100
BACKING_CHANNELS = 2

# Bypassed values for the two Effects (shared/adjustments.ts's REVERB_AMOUNT_MIN
# and LOWPASS_HZ_MAX). At these values a filter is left out of the graph
# entirely rather than merely configured to be neutral (ticket 07).
REVERB_AMOUNT_BYPASSED = 0
LOWPASS_HZ_BYPASSED = 20000

# Which side of a Mix the Effects colour (shared/adjustments.ts's
# EFFECTS_TARGETS). "backing" is the default, and the only behaviour there
# was before the target existed.
EFFECTS_TARGET_BACKING = "backing"
EFFECTS_TARGETS_ON_BACKING = frozenset({EFFECTS_TARGET_BACKING, "both"})
EFFECTS_TARGETS_ON_VOCAL = frozenset({"vocal", "both"})

# The one impulse response ADR 0003 requires both engines to share; the
# browser's `app/audio/engine.ts` loads the same file from `public/audio/`.
IMPULSE_RESPONSE_PATH = Path(__file__).parent / "assets" / "large-hall-ir.wav"


class AudioError(RuntimeError):
    """ffmpeg or ffprobe refused the input. The message carries the tool's own diagnostics."""


def normalize_to_backing_track(src: Path, dst: Path) -> None:
    """Decode any Source audio and write it as the 44.1 kHz stereo 16-bit WAV Backing Track."""
    dst.parent.mkdir(parents=True, exist_ok=True)
    tmp = dst.with_suffix(".part.wav")
    result = subprocess.run(
        [
            "ffmpeg",
            "-y",
            "-nostdin",
            "-hide_banner",
            "-loglevel",
            "error",
            "-i",
            str(src),
            "-vn",
            "-ac",
            str(BACKING_CHANNELS),
            "-ar",
            str(BACKING_SAMPLE_RATE),
            "-c:a",
            "pcm_s16le",
            str(tmp),
        ],
        capture_output=True,
        text=True,
    )
    if result.returncode != 0:
        tmp.unlink(missing_ok=True)
        detail = _clean_ffmpeg_stderr(result.stderr) or f"exit code {result.returncode}"
        raise AudioError(f"ffmpeg could not decode {src.name}: {detail}")
    tmp.replace(dst)


_CONTEXT_PREFIX = re.compile(r"^\[[^\]]*\]\s*")


def _clean_ffmpeg_stderr(stderr: str) -> str:
    """Drop ffmpeg's `[in#0 @ 0x...]` context tags; a singer reading the card does not need them."""
    lines = [_CONTEXT_PREFIX.sub("", line).strip() for line in stderr.splitlines()]
    return "\n".join(line for line in lines if line)


def _append_effects(
    segments: list[str],
    *,
    label: str,
    prefix: str,
    reverb_amount: int,
    lowpass_hz: int,
    impulse_label: str,
) -> str:
    """Appends the Effects — reverb, then low-pass, the order both engines build
    them in (ADR 0003) — onto one signal's `segments`, and returns the label
    they leave it on.

    Either filter at its bypassed value contributes no segment at all, so a
    signal the Effects Target does not reach passes through exactly as it would
    have before there were Effects to apply.
    """
    if reverb_amount > REVERB_AMOUNT_BYPASSED:
        wet = reverb_amount / 100
        dry = 1 - wet
        segments.append(f"[{label}]asplit=2[{prefix}_dry][{prefix}_wet_in]")
        segments.append(f"[{prefix}_wet_in]{impulse_label}afir=dry=1:wet=1[{prefix}_wet]")
        segments.append(
            f"[{prefix}_dry][{prefix}_wet]amix=inputs=2:"
            f"weights={dry:.6f} {wet:.6f}:normalize=0[{prefix}_reverbed]"
        )
        label = f"{prefix}_reverbed"

    if lowpass_hz < LOWPASS_HZ_BYPASSED:
        segments.append(f"[{label}]lowpass=f={lowpass_hz}[{prefix}_lp]")
        label = f"{prefix}_lp"

    return label


def render_mix(
    *,
    backing: Path,
    vocal: Path,
    dst_mp3: Path,
    dst_wav: Path | None,
    tempo: float,
    pitch: float,
    vocal_wall_ms: float,
    vocal_gain: float,
    backing_gain: float,
    target_duration_ms: int,
    reverb_amount: int = REVERB_AMOUNT_BYPASSED,
    lowpass_hz: int = LOWPASS_HZ_BYPASSED,
    effects_target: str = EFFECTS_TARGET_BACKING,
) -> None:
    """Renders a Mix: the Backing Track stretched by Rubber Band, then the two
    Effects (ticket 07), then the Take's dry vocal placed at `vocal_wall_ms`
    (already converted from song time to wall time by the caller) and scaled
    by `vocal_gain`, summed with the backing scaled by `backing_gain`. The
    result always covers exactly `target_duration_ms` — the Backing Track's
    own duration is padded or trimmed to it, since Rubber Band's extreme pitch
    shifts land a few percent short of the input length on their own
    (ADR 0003, ADR 0004).

    Reverb is `afir` convolved against the bundled impulse response, always
    fully wet on its own output; the dry/wet crossfade the singer actually
    hears is an `amix` of that against an unprocessed copy, mirroring the
    convolver/dry-gain/wet-gain graph `app/audio/effects-chain.ts` builds in the
    browser. The low-pass is ffmpeg's `lowpass` filter at its default (two-pole)
    Q, matching a Web Audio `BiquadFilterNode` lowpass at its own default
    without tuning either side. Both filters are left out of the filter graph
    entirely at their bypassed values, not merely configured to be neutral, so
    a Mix rendered from an untouched Track is what phase one would have
    produced.

    `effects_target` decides which of the two signals that pair of filters
    lands on (ticket 13): the vocal, the backing, both, or neither. Each side
    builds the same two segments in the same order, so a Mix reproduces
    whichever chains the Review screen had active. The vocal's `volume` stays
    ahead of its Effects rather than after them as the backing's is; both
    filters are linear, so a scalar commutes through them exactly, and leaving
    it there keeps the default target's filter graph the string it has always
    been.

    A negative `vocal_wall_ms` means the Take started before the tempo-adjusted
    Backing Track does (a large negative nudge on an early Take); rather than
    delaying, that much is trimmed off the front of the vocal so it still lands
    at wall time zero.
    """
    dst_mp3.parent.mkdir(parents=True, exist_ok=True)
    target_seconds = target_duration_ms / 1000

    if vocal_wall_ms >= 0:
        delay_ms = round(vocal_wall_ms)
        vocal_chain = f"adelay={delay_ms}|{delay_ms},volume={vocal_gain}"
    else:
        trim_seconds = -vocal_wall_ms / 1000
        vocal_chain = f"atrim=start={trim_seconds:.6f},asetpts=PTS-STARTPTS,volume={vocal_gain}"

    on_backing = effects_target in EFFECTS_TARGETS_ON_BACKING
    on_vocal = effects_target in EFFECTS_TARGETS_ON_VOCAL
    backing_reverb = reverb_amount if on_backing else REVERB_AMOUNT_BYPASSED
    vocal_reverb = reverb_amount if on_vocal else REVERB_AMOUNT_BYPASSED

    # One impulse response input feeds at most two `afir`s, and a stream can be
    # consumed only once, so it is split when both sides want it.
    backing_wants_ir = backing_reverb > REVERB_AMOUNT_BYPASSED
    vocal_wants_ir = vocal_reverb > REVERB_AMOUNT_BYPASSED
    ir_segments: list[str] = []
    backing_ir = vocal_ir = "[2:a]"
    if backing_wants_ir and vocal_wants_ir:
        ir_segments.append("[2:a]asplit=2[ir_bg][ir_voc]")
        backing_ir, vocal_ir = "[ir_bg]", "[ir_voc]"

    backing_segments = [f"[0:a]rubberband=tempo={tempo}:pitch={pitch}[stretched]"]
    backing_label = _append_effects(
        backing_segments,
        label="stretched",
        prefix="bg",
        reverb_amount=backing_reverb,
        lowpass_hz=lowpass_hz if on_backing else LOWPASS_HZ_BYPASSED,
        impulse_label=backing_ir,
    )
    backing_segments.append(
        f"[{backing_label}]apad=whole_dur={target_seconds:.6f},atrim=end={target_seconds:.6f},"
        f"asetpts=PTS-STARTPTS,volume={backing_gain}[bg]"
    )

    vocal_segments = [f"[1:a]{vocal_chain}[voc]"]
    vocal_label = _append_effects(
        vocal_segments,
        label="voc",
        prefix="voc",
        reverb_amount=vocal_reverb,
        lowpass_hz=lowpass_hz if on_vocal else LOWPASS_HZ_BYPASSED,
        impulse_label=vocal_ir,
    )

    filter_complex = ";".join(
        [
            *ir_segments,
            *backing_segments,
            *vocal_segments,
            f"[bg][{vocal_label}]amix=inputs=2:duration=first:normalize=0[out]",
        ]
    )

    inputs = ["-i", str(backing), "-i", str(vocal)]
    if backing_wants_ir or vocal_wants_ir:
        inputs += ["-i", str(IMPULSE_RESPONSE_PATH)]

    tmp_wav = dst_mp3.with_suffix(".part.wav")
    result = subprocess.run(
        [
            "ffmpeg",
            "-y",
            "-nostdin",
            "-hide_banner",
            "-loglevel",
            "error",
            *inputs,
            "-filter_complex",
            filter_complex,
            "-map",
            "[out]",
            "-ar",
            str(BACKING_SAMPLE_RATE),
            "-ac",
            str(BACKING_CHANNELS),
            "-c:a",
            "pcm_s16le",
            str(tmp_wav),
        ],
        capture_output=True,
        text=True,
    )
    if result.returncode != 0:
        tmp_wav.unlink(missing_ok=True)
        detail = _clean_ffmpeg_stderr(result.stderr) or f"exit code {result.returncode}"
        raise AudioError(f"ffmpeg could not render the Mix: {detail}")

    tmp_mp3 = dst_mp3.with_suffix(".part.mp3")
    mp3_result = subprocess.run(
        [
            "ffmpeg",
            "-y",
            "-nostdin",
            "-hide_banner",
            "-loglevel",
            "error",
            "-i",
            str(tmp_wav),
            "-c:a",
            "libmp3lame",
            "-b:a",
            "320k",
            str(tmp_mp3),
        ],
        capture_output=True,
        text=True,
    )
    if mp3_result.returncode != 0:
        tmp_wav.unlink(missing_ok=True)
        tmp_mp3.unlink(missing_ok=True)
        detail = _clean_ffmpeg_stderr(mp3_result.stderr) or f"exit code {mp3_result.returncode}"
        raise AudioError(f"ffmpeg could not encode the Mix to MP3: {detail}")
    tmp_mp3.replace(dst_mp3)

    if dst_wav is not None:
        tmp_wav.replace(dst_wav)
    else:
        tmp_wav.unlink(missing_ok=True)


def probe_duration_ms(path: Path) -> int:
    result = subprocess.run(
        [
            "ffprobe",
            "-v",
            "error",
            "-show_entries",
            "format=duration",
            "-of",
            "json",
            str(path),
        ],
        capture_output=True,
        text=True,
    )
    if result.returncode != 0:
        raise AudioError(f"ffprobe could not read {path.name}: {result.stderr.strip()}")
    try:
        seconds = float(json.loads(result.stdout)["format"]["duration"])
    except (KeyError, ValueError, json.JSONDecodeError) as exc:
        raise AudioError(f"ffprobe reported no duration for {path.name}") from exc
    return round(seconds * 1000)

File: server/lib/audio.ts
import { existsSync } from 'node:fs'
import { mkdir, rename, rm } from 'node:fs/promises'
import { dirname, resolve } from 'node:path'
import { spawn } from 'node:child_process'
import {
  effectsReachBacking,
  effectsReachVocal,
  LOWPASS_HZ_MAX,
  REVERB_AMOUNT_MIN,
  type EffectsTarget,
} from '../../shared/adjustments'

/**
 * Thin wrappers over ffmpeg and ffprobe, ported from `worker/akapela_worker/audio.py`
 * (ticket 03/04 of `.scratch/worker-to-typescript/`; ffmpeg was always just a
 * subprocess call, never a Python-specific dependency).
 *
 * Every stored audio master is 44.1 kHz stereo WAV (ADR 0005).
 */

export const BACKING_SAMPLE_RATE = 44100
export const BACKING_CHANNELS = 2

export class AudioError extends Error {}

/** Runs ffmpeg/ffprobe and rejects with `AudioError` on a non-zero exit. */
function run(bin: 'ffmpeg' | 'ffprobe', args: string[]): Promise<string> {
  return new Promise((resolve, reject) => {
    const child = spawn(bin, args)
    let stdout = ''
    let stderr = ''
    child.stdout.on('data', d => (stdout += d))
    child.stderr.on('data', d => (stderr += d))
    child.on('error', error => reject(new AudioError(`could not start ${bin}: ${error.message}`)))
    child.on('close', (code) => {
      if (code === 0) resolve(stdout)
      else reject(new AudioError(`${bin} exited ${code}: ${cleanFfmpegStderr(stderr) || `exit code ${code}`}`))
    })
  })
}

/** Drops ffmpeg's `[in#0 @ 0x...]` context tags; a singer reading the card does not need them. */
function cleanFfmpegStderr(stderr: string): string {
  return stderr
    .split('\n')
    .map(line => line.replace(/^\[[^\]]*\]\s*/, '').trim())
    .filter(Boolean)
    .join('\n')
}

/** Decodes any Source audio and writes it as the 44.1 kHz stereo 16-bit WAV Backing Track. */
export async function normalizeToBackingTrack(src: string, dst: string): Promise<void> {
  await mkdir(dirname(dst), { recursive: true })
  const tmp = dst.replace(/\.wav$/, '.part.wav')
  try {
    await run('ffmpeg', [
      '-y',
      '-nostdin',
      '-hide_banner',
      '-loglevel',
      'error',
      '-i',
      src,
      '-vn',
      '-ac',
      String(BACKING_CHANNELS),
      '-ar',
      String(BACKING_SAMPLE_RATE),
      '-c:a',
      'pcm_s16le',
      tmp,
    ])
  }
  catch (error) {
    await rm(tmp, { force: true })
    throw error instanceof AudioError
      ? new AudioError(`ffmpeg could not decode ${src}: ${error.message.replace(/^ffmpeg exited \d+: /, '')}`)
      : error
  }
  await rename(tmp, dst)
}

/** The duration of an audio file, in milliseconds. */
export async function probeDurationMs(path: string): Promise<number> {
  let stdout: string
  try {
    stdout = await run('ffprobe', [
      '-v',
      'error',
      '-show_entries',
      'format=duration',
      '-of',
      'json',
      path,
    ])
  }
  catch (error) {
    throw error instanceof AudioError
      ? new AudioError(`ffprobe could not read ${path}: ${error.message.replace(/^ffprobe exited \d+: /, '')}`)
      : error
  }
  const seconds = Number(JSON.parse(stdout)?.format?.duration)
  if (!Number.isFinite(seconds)) throw new AudioError(`ffprobe reported no duration for ${path}`)
  return Math.round(seconds * 1000)
}

/**
 * The one impulse response ADR 0003 requires both engines to share; the
 * browser's `app/audio/engine.ts` loads the same file from `public/audio/`.
 * Resolved by checking the built output first rather than assuming a `cwd`:
 * `pnpm dev` serves straight from `public/`, the compose image serves the
 * copy Nuxt's build already puts in `.output/public/` next to it.
 */
function impulseResponsePath(): string {
  const built = resolve('.output/public/audio/large-hall-ir.wav')
  return existsSync(built) ? built : resolve('public/audio/large-hall-ir.wav')
}

export interface RenderMixOptions {
  backing: string
  vocal: string
  dstMp3: string
  dstWav: string | null
  tempo: number
  pitch: number
  vocalWallMs: number
  vocalGain: number
  backingGain: number
  targetDurationMs: number
  reverbAmount?: number
  lowpassHz?: number
  effectsTarget?: EffectsTarget
}

/**
 * Appends the Effects — reverb, then low-pass, the order both engines build
 * them in (ADR 0003) — onto one signal's `segments`, and returns the label
 * they leave it on.
 *
 * Either filter at its bypassed value contributes no segment at all, so a
 * signal the Effects Target does not reach passes through exactly as it would
 * have before there were Effects to apply.
 */
export function appendEffects(
  segments: string[],
  options: { label: string, prefix: string, reverbAmount: number, lowpassHz: number, impulseLabel: string },
): string {
  let { label } = options
  const { prefix, reverbAmount, lowpassHz, impulseLabel } = options

  if (reverbAmount > REVERB_AMOUNT_MIN) {
    const wet = reverbAmount / 100
    const dry = 1 - wet
    segments.push(`[${label}]asplit=2[${prefix}_dry][${prefix}_wet_in]`)
    segments.push(`[${prefix}_wet_in]${impulseLabel}afir=dry=1:wet=1[${prefix}_wet]`)
    segments.push(
      `[${prefix}_dry][${prefix}_wet]amix=inputs=2:`
      + `weights=${dry.toFixed(6)} ${wet.toFixed(6)}:normalize=0[${prefix}_reverbed]`,
    )
    label = `${prefix}_reverbed`
  }

  if (lowpassHz < LOWPASS_HZ_MAX) {
    segments.push(`[${label}]lowpass=f=${lowpassHz}[${prefix}_lp]`)
    label = `${prefix}_lp`
  }

  return label
}

/**
 * Renders a Mix: the Backing Track stretched by Rubber Band, then the two
 * Effects, then the Take's dry vocal placed at `vocalWallMs` (already
 * converted from song time to wall time by the caller) and scaled by
 * `vocalGain`, summed with the backing scaled by `backingGain`. The result
 * always covers exactly `targetDurationMs` — the Backing Track's own duration
 * is padded or trimmed to it, since Rubber Band's extreme pitch shifts land a
 * few percent short of the input length on their own (ADR 0003, ADR 0004).
 *
 * Ported from `worker/akapela_worker/audio.py`'s `render_mix` (ticket 04).
 */
export async function renderMix(options: RenderMixOptions): Promise<void> {
  const {
    backing, vocal, dstMp3, dstWav, tempo, pitch, vocalWallMs, vocalGain, backingGain, targetDurationMs,
    reverbAmount = REVERB_AMOUNT_MIN, lowpassHz = LOWPASS_HZ_MAX, effectsTarget = 'backing',
  } = options

  await mkdir(dirname(dstMp3), { recursive: true })
  const targetSeconds = targetDurationMs / 1000

  let vocalChain: string
  if (vocalWallMs >= 0) {
    const delayMs = Math.round(vocalWallMs)
    vocalChain = `adelay=${delayMs}|${delayMs},volume=${vocalGain}`
  }
  else {
    const trimSeconds = -vocalWallMs / 1000
    vocalChain = `atrim=start=${trimSeconds.toFixed(6)},asetpts=PTS-STARTPTS,volume=${vocalGain}`
  }

  const onBacking = effectsReachBacking(effectsTarget)
  const onVocal = effectsReachVocal(effectsTarget)
  const backingReverb = onBacking ? reverbAmount : REVERB_AMOUNT_MIN
  const vocalReverb = onVocal ? reverbAmount : REVERB_AMOUNT_MIN

  // One impulse response input feeds at most two `afir`s, and a stream can be
  // consumed only once, so it is split when both sides want it.
  const backingWantsIr = backingReverb > REVERB_AMOUNT_MIN
  const vocalWantsIr = vocalReverb > REVERB_AMOUNT_MIN
  const irSegments: string[] = []
  let backingIr = '[2:a]'
  let vocalIr = '[2:a]'
  if (backingWantsIr && vocalWantsIr) {
    irSegments.push('[2:a]asplit=2[ir_bg][ir_voc]')
    backingIr = '[ir_bg]'
    vocalIr = '[ir_voc]'
  }

  const backingSegments = [`[0:a]rubberband=tempo=${tempo}:pitch=${pitch}[stretched]`]
  const backingLabel = appendEffects(backingSegments, {
    label: 'stretched',
    prefix: 'bg',
    reverbAmount: backingReverb,
    lowpassHz: onBacking ? lowpassHz : LOWPASS_HZ_MAX,
    impulseLabel: backingIr,
  })
  backingSegments.push(
    `[${backingLabel}]apad=whole_dur=${targetSeconds.toFixed(6)},atrim=end=${targetSeconds.toFixed(6)},`
    + `asetpts=PTS-STARTPTS,volume=${backingGain}[bg]`,
  )

  const vocalSegments = [`[1:a]${vocalChain}[voc]`]
  const vocalLabel = appendEffects(vocalSegments, {
    label: 'voc',
    prefix: 'voc',
    reverbAmount: vocalReverb,
    lowpassHz: onVocal ? lowpassHz : LOWPASS_HZ_MAX,
    impulseLabel: vocalIr,
  })

  const filterComplex = [
    ...irSegments,
    ...backingSegments,
    ...vocalSegments,
    `[bg][${vocalLabel}]amix=inputs=2:duration=first:normalize=0[out]`,
  ].join(';')

  const inputs = ['-i', backing, '-i', vocal]
  if (backingWantsIr || vocalWantsIr) inputs.push('-i', impulseResponsePath())

  const tmpWav = dstMp3.replace(/\.mp3$/, '.part.wav')
  try {
    await run('ffmpeg', [
      '-y', '-nostdin', '-hide_banner', '-loglevel', 'error',
      ...inputs,
      '-filter_complex', filterComplex,
      '-map', '[out]',
      '-ar', String(BACKING_SAMPLE_RATE),
      '-ac', String(BACKING_CHANNELS),
      '-c:a', 'pcm_s16le',
      tmpWav,
    ])
  }
  catch (error) {
    await rm(tmpWav, { force: true })
    throw error instanceof AudioError
      ? new AudioError(`ffmpeg could not render the Mix: ${error.message.replace(/^ffmpeg exited \d+: /, '')}`)
      : error
  }

  const tmpMp3 = dstMp3.replace(/\.mp3$/, '.part.mp3')
  try {
    await run('ffmpeg', [
      '-y', '-nostdin', '-hide_banner', '-loglevel', 'error',
      '-i', tmpWav,
      '-c:a', 'libmp3lame',
      '-b:a', '320k',
      tmpMp3,
    ])
  }
  catch (error) {
    await rm(tmpWav, { force: true })
    await rm(tmpMp3, { force: true })
    throw error instanceof AudioError
      ? new AudioError(`ffmpeg could not encode the Mix to MP3: ${error.message.replace(/^ffmpeg exited \d+: /, '')}`)
      : error
  }
  await rename(tmpMp3, dstMp3)

  if (dstWav !== null) await rename(tmpWav, dstWav)
  else await rm(tmpWav, { force: true })
}
